backend/src/main/kotlin/demo/acme/controller/StoreController.kt

```diff
@@ -0,0 +1,80 @@
+package demo.acme.controller
+
+import com.beust.klaxon.Klaxon
+import demo.acme.StoreRepository
+import demo.acme.component.UpdateStoreInformationSchedule
+import demo.acme.model.Store
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.core.io.InputStreamResource
+import org.springframework.data.domain.PageRequest
+import org.springframework.data.domain.Sort
+import org.springframework.http.HttpStatus
+import org.springframework.http.MediaType
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+import org.springframework.web.server.ResponseStatusException
+import java.io.File
+import java.io.FileInputStream
+
+
+@RestController
+class StoreController {
+    @Autowired
+    private lateinit var updateStoreInformationSchedule: UpdateStoreInformationSchedule
+
+    @Autowired
+    private lateinit var repository: StoreRepository
+
+    @GetMapping("/api/getStores", produces = [MediaType.APPLICATION_JSON_VALUE])
+    fun getStores(@RequestParam page: Int, @RequestParam(required = false) storeNameFilter: String?): String {
+        val pageSize = 20
+        val pageRequest = PageRequest.of(page, pageSize, Sort.by("id"))
+        val storesPage = if (storeNameFilter == null || storeNameFilter.isEmpty())
+            repository.findAll(pageRequest)
+        else repository.findByNameContaining(storeNameFilter, pageRequest)
+        return Klaxon().toJsonString(
+            object {
+                val storesPage = storesPage.get().toList()
+                val totalPages = storesPage.totalPages
+            })
+    }
+
+    @PutMapping("/api/updateStore")
+    fun updateStore(@RequestBody storeAsJson: String): String {
```

Comment:
Have you tried this instead of using Klaxon?

`fun updateStore(@RequestBody store: Store): String {`

Also, in your architecture design you are using the Store persistence class in the controller layer (persistence and controllers are not decoupled). Can you name a few disadvantages of this design decision?

Reply:
*facepalm*. And it automatically handles bad requests too. Had no idea, thanks :)

As for the question, I suppose tutorials and guides usually lead us to insert business logic in controllers seems wrong. Controllers should simply serve as bridges between the model and views. Refactoring logic out into services would ease the maintenance and facilitate expansion and testing without major changes to the controller. Also we'd be less prone to code duplication. Perhaps an issue with mixing DB entities and the controller might be exposing too much data by accident. There's nothing sensitive about the Store entity atm, but there could be.